File: 2597. The Number of Beautiful Subsets.java
class Solution {
    public int beautifulSubsets(int[] nums, int k) {
        int maxValue = Integer.MIN_VALUE;
        for (int i : nums) maxValue = Math.max(maxValue, i);
        int[] count = new int[maxValue+1];
        return find(nums, k, count, 0)-1;
    }
    private int find(int[] nums, int k, int[] count, int index) {
        if (index >= nums.length) return 1;
        int answer = find(nums, k, count, index+1);
        if ((nums[index]-k < 0 || count[nums[index]-k] == 0)
            && (nums[index]+k >= count.length || count[nums[index]+k] == 0)) {
            count[nums[index]]++;
            answer += find(nums, k, count, index+1);
            count[nums[index]]--;
        }
        return answer;
    }
}

File: 2597. The Number of Beautiful Subsets.go
func beautifulSubsets(nums []int, k int) int {
	return find(nums, make([]int, 1001), k, 0) - 1
}

func find(nums, count []int, k, index int) int {
	if index >= len(nums) {
		return 1
	}
	answer := find(nums, count, k, index+1)
	condition := nums[index]+k >= len(count) || count[nums[index]+k] == 0
	condition = condition && (nums[index]-k < 0 || count[nums[index]-k] == 0)
	if condition {
		count[nums[index]] += 1
		answer += find(nums, count, k, index+1)
		count[nums[index]] -= 1
	}
	return answer
}
